### PR TITLE
Fix immediate execution of approved scheduled jobs

### LINKED ISSUE #7038: Run Immediate Job once approved is stuck when Scheduled
Sorry about the title, unsure how to summarise this problem into a brief sentence!

### Environment
* Nautobot version (Docker tag too if applicable): v2.4.4
* Python version: 3.11
* Database platform, version: Postgres
* Middleware(s): Redis

### Steps to Reproduce
1. Create a Job, make this Job Approval Required
2. Run the Job, set it to be "run immediately"
3. Approve the Job, no pop up will occur even though the Job scheduled run time is in the past
4. Observe that the Job is set to run in the past, as the run time is determined when the original request to run the job is created
5. Observe that the Job is "one off" false
6. The job is stuck in Scheduled Jobs and will never run

### Expected Behavior
A job that is sent for approval that is set to be "run immediately" should be correctly flagged as one off, so that the pop up is received when approval is attempted, and ran once approved.

Additionally, if run immediately, the job should ideally not be scheduled, but instead triggered to be ran as it is successfully approved (this would somewhat mimic the dry run logic of a job to be approved, where it is enqueued). However, if scheduled, it should be scheduled in such a way that it will run once approved.

### Observed Behavior
Job is approved with no pop up even though it is in the past, and it is stuck in the Scheduled jobs queue. This results in Jobs that follow the process of Creation -> Approval -> Approve -> Scheduled, when set to be "Run Immeditately" are never actually ran and lost.

Additionally, it also looks to always show an empty job queue value in the approval screen for the job, even if a default job queue is selected or a job queue value is provided in the jobs run creation.

Note: Celery-beat is confirmed to be on, and other Scheduled Jobs are being triggered daily while this problem is being observed.

The requirement for "one off" for the pop up about job time being in the past can be seen here;
https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/api/views.py#L965

Courtesy of @joewesch for also highlighting that one_off seems to be restricted to type future here: https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/models/jobs.py#L1375

### PR DESCRIPTION
**Pull request by dsoftait**
# Closes #7038 
# What's Changed
Fixed issue where approved scheduled jobs set to run "immediately" were not executed, by changing the `create_schedule` method in `ScheduledJob`.

# TODO
- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/views.py [hunk 1]
```diff
@@ -1687,6 +1687,15 @@ def post(self, request, pk):
...
                 scheduled_job.approved_at = timezone.now()
+                if (
+                    scheduled_job.interval == JobExecutionType.TYPE_IMMEDIATELY
+                    and scheduled_job.start_time < timezone.localtime()
+                ):
+                    # This is needed because otherwise multiple jobs will be scheduled, not just one.
+                    scheduled_job.one_off = True
+                    # This is needed because if we don't change the interval to TYPE_FUTURE, the scheduler will not send this task.
+                    scheduled_job.interval = JobExecutionType.TYPE_FUTURE
+                    scheduled_job.start_time = timezone.localtime()
```
Comment (dsoftait) [inline nautobot/extras/views.py]: Should I also add this change [here](https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/api/views.py#L972)?

File: nautobot/extras/views.py [hunk 2]
```diff
@@ -1687,6 +1687,15 @@ def post(self, request, pk):
                     return render(request, "extras/job_approval_confirmation.html", {"scheduled_job": scheduled_job})
                 scheduled_job.approved_by_user = request.user
                 scheduled_job.approved_at = timezone.now()
+                if (
+                    scheduled_job.interval == JobExecutionType.TYPE_IMMEDIATELY
+                    and scheduled_job.start_time < timezone.localtime()
```
Comment (dsoftait) [inline nautobot/extras/views.py]: when `scheduled_job.interval == JobExecutionType.TYPE_IMMEDIATELY` then `scheduled_job.start_time < timezone.localtime()` is probably always true, so I think this check is redundant.

File: nautobot/extras/views.py [hunk 3]
```diff
@@ -1687,6 +1687,15 @@ def post(self, request, pk):
...
+                ):
+                    # This is needed because otherwise multiple jobs will be scheduled, not just one.
+                    scheduled_job.one_off = True
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: Should this have been set in `create_schedule()`? Right now in that function we set `one_off=(interval == JobExecutionType.TYPE_FUTURE)` and I wonder if that's incorrect.

File: nautobot/extras/views.py [hunk 4]
```diff
@@ -1687,6 +1687,15 @@ def post(self, request, pk):
...
+                    # This is needed because if we don't change the interval to TYPE_FUTURE, the scheduler will not send this task.
+                    scheduled_job.interval = JobExecutionType.TYPE_FUTURE
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: Given that `TYPE_IMMEDIATELY` isn't one of the values in `JobExecutionType.SCHEDULE_CHOICES`, I'm actually thinking the intent maybe was that we should never create ScheduledJobs with an interval of `TYPE_IMMEDIATELY`? Wondering if a "more correct" fix would be in `JobRunView.post()` to catch the special case of:

- User requested TYPE_IMMEDIATELY
- and this is not a dry-run
- but the job requires approval so it can't actually be executed immediately

and apply these fixups there (or perhaps more generally in `ScheduledJob.create_schedule()`?) rather than after-the-fact here.


Also, let's make sure a similar fix applies in the REST API as well, whichever approach we go with.

Comment (dsoftait) [reply nautobot/extras/views.py]: I understand. After analyzing what you wrote, changing `create_schedule` makes more sense to me than modifying `JobRunView.post()`. So I'll go in that direction. Thanks for your feedback.

Comment (dsoftait) [reply nautobot/extras/views.py]: done

**Comment by dsoftait** [comment]
> Looks good to me - I assume you've tested the rework manually since we don't have much/any automated test coverage of scheduled job execution?

Manually too, but actually we have some:
api:
- https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/tests/test_api.py#L1625
- https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/tests/test_api.py#L1854
- https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/tests/test_api.py#L1881
- https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/tests/test_api.py#L2681
- https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/tests/test_api.py#L2688
- https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/tests/test_api.py#L2705

The last three tests cover the approval workflow. Therefore, if a scheduled job uses the `TYPE_FUTURE` interval instead of `TYPE_IMMEDIATELY` (which is specifically tested in tests below), it is already accounted for in these tests.
- https://github.com/nautobot/nautobot/pull/7414/files#diff-3444a4f7066243e26b2dd66ba2080c48e0cb75f5bf217b1f397aeb56c5997431R2167
- https://github.com/nautobot/nautobot/pull/7414/files#diff-63390a9316fbabcd9c87d7ea0980d9ff2738d5da7d7def31e8f152d5fcf92996R3095

views:
- https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/tests/test_views.py#L2980
- https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/tests/test_views.py#L2411

But if you think I should add something else, let me know.

**Comment by glennmatthews** [comment]
My point was more that most of those tests don't actually *execute* the scheduled job, just create the database record, hence why none of them caught the originally reported issue.

**Comment by dsoftait** [comment]
> My point was more that most of those tests don't actually _execute_ the scheduled job, just create the database record, hence why none of them caught the originally reported issue.

I see, I checked it and as I understand it, in order to do this, we need to fix the problem with `NautobotDatabaseScheduler` that was reported [here](https://github.com/nautobot/nautobot/pull/6517/files#diff-3444a4f7066243e26b2dd66ba2080c48e0cb75f5bf217b1f397aeb56c5997431R2141), because the approved scheduled job creates a `publish_event`, and then celery-beat is responsible for the rest of the flow. 

I created a ticket for this https://github.com/nautobot/nautobot/issues/7441
